### PR TITLE
Fix srtp constraint printing

### LINKED ISSUE #5768: SRTP reports a required constraint (twice) that has already been specified
Type resolution reports FS0001 and FS0043 with the same error, that a constraint is required - but that constraint is already specified

#### Repro steps

Crazy code sample:
```fsharp
type Tuple =
    static member inline TupleMap ((a, b),          _ : Tuple) = fun f -> (f a, f b)
    static member inline TupleMap ((a, b, c),       _ : Tuple) = fun f -> (f a, f b, f c)
    static member inline TupleMap ((a, b, c, d),    _ : Tuple) = fun f -> (f a, f b, f c, f d)
    
    static member inline Map f (x: 'Tin) : 'Tout =
        let inline call_2 (a: ^a, b : ^b) = ((^a or ^b) : (static member TupleMap : ^b * _ -> ^t) (b, a))
        call_2 (Unchecked.defaultof<Tuple>, x) f

type IOption<'T> =
    abstract member Value : 'T

module Something = 
    let inline tupleMap f x = Tuple.Map f x

    let inline addOptionValues< ^value, ^options, ^values, 'item when 
                                  'item :> IOption< ^value>
                                  and (Tuple or ^options) : (static member TupleMap : ^options * Tuple -> ('item -> ^value) -> ^values)>
                                  (addUp : ^values -> ^value, sourceOptions : ^options) =
        let getValue (i : 'item) = i.Value
        let allValues : ^values = tupleMap getValue sourceOptions
        let result : ^value = addUp allValues
        result
``` 

The line 
```fsharp
        let allValues : ^values = tupleMap getValue sourceOptions
```
Gives the errors
```
Error	FS0043	A type parameter is missing a constraint 'when (Tuple or  ^options) : (static member TupleMap :  ^options * Tuple -> ('item ->  ^value) ->  ^values)'	
Error	FS0001	A type parameter is missing a constraint 'when (Tuple or  ^options) : (static member TupleMap :  ^options * Tuple -> ('item ->  ^value) ->  ^values)'	
```

#### Related information 

Provide any related information 

* Windows 10
* Visual F# Tools 10.2 for F# 4.5   15.8.0.0.  Commit Hash: 55a3dc3231c95c77f81ee53f7d29152029da7408.
* Microsoft Visual F# Tools 10.2 for F# 4.5

### PR DESCRIPTION
**Pull request by abelbraaksma**
Fixes #5768

TLDR: srtp contraints returning a function did not get pretty-printed for overloads-suggestions like `'a -> 'b -> ('x -> 'y)`, instead, you'd see `'a -> 'b -> 'x -> 'y`. That is, the parens were omitted if a function was returned. Same is true for type or constraint missing errors. 

This continues https://github.com/dotnet/fsharp/pull/5948. It was 2 years behind master, and creating a new PR seemed a simpler way to get this done. The changes w.r.t. that PR are small:

* Renamed the new test `neg111` to `neg112` and
* Renamed `neg111.fs` to `neg112.fs` and
* Renamed `neg111.bsl` to `neg112.bsl`

Let's see what CI makes of this now ;).

@realvictorprm, this was originally your PR and I branched off your branch to keep the history intact. I hope it's ok that I try to continue & finish it ;). Since I cannot make changes to an existing PR, I decided to make a new one. I don't claim to know all details of your fix, but from the look of it, it seemed quite straight-forward and it was already reviewed & accepted as-is.

### CODE REVIEW

File: src/fsharp/NicePrint.fs [hunk 1]
```diff
@@ -892,7 +892,7 @@ module private PrintTypes =
                 | tys -> bracketL (layoutTypesWithInfoAndPrec denv env 2 (wordL (tagKeyword "or")) tys)
             tysL ^^ wordL (tagPunctuation ":") ---
                 bracketL (stat ++ wordL (tagMember nm) ^^ wordL (tagPunctuation ":") ---
-                        ((layoutTypesWithInfoAndPrec denv env 2 (wordL (tagPunctuation "*")) argtys --- wordL (tagPunctuation "->")) --- (layoutTypeWithInfo denv env rty)))
```
Comment (cartermp) [inline src/fsharp/NicePrint.fs:895]: Could probably be simplified to `layoutReturnType`

**Comment by abelbraaksma** [comment]
Hmm, I'm now suddenly locally getting this:

```
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(14,12): error CS0579: Duplicate 'System.Reflection.AssemblyCompanyAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(15,12): error CS0579: Duplicate 'System.Reflection.AssemblyConfigurationAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(17,12): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(18,12): error CS0579: Duplicate 'System.Reflection.AssemblyInformationalVersionAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(19,12): error CS0579: Duplicate 'System.Reflection.AssemblyProductAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(20,12): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
D:\Projects\OpenSource\FSharp\artifacts\obj\PEVerify\Debug\net472\PEVerify.AssemblyInfo.cs(21,12): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute [D:\Projects\OpenSource\FSharp\tests\fsharpqa\testenv\src\PEVerify\PEVerify.csproj]
```

Haven't seen it before in this solution. I understand the error, but the file is auto-generated. Wonder if it has something to do with me merging 2 years of stuff. Hope CI doesn't see this ;).

**Comment by abelbraaksma** [comment]
Some errors like the following, but these were expected, as these tests were added _after_ this change and are related to the stringification of constraints. I'll look into them

```diff
-  - static member Neg117.Superpower.Transformer.Transform :  ^f * Neg117.TargetB.TargetB * Neg117.Superpower.Transformer -> (Neg117.TargetB.TransformerKind ->  ^f) when (Neg117.TargetB.TargetB or  ^f) : (static member Transform :  ^f * Neg117.TargetB.TargetB -> Neg117.TargetB.TransformerKind ->  ^f) // Argument at index 1 doesn't match
+  - static member Neg117.Superpower.Transformer.Transform :  ^f * Neg117.TargetB.TargetB * Neg117.Superpower.Transformer -> (Neg117.TargetB.TransformerKind ->  ^f) when (Neg117.TargetB.TargetB or  ^f) : (static member Transform :  ^f * Neg117.TargetB.TargetB -> (Neg117.TargetB.TransformerKind ->  ^f)) // Argument at index 1 doesn't match
diff between [D:\a\1\s\tests\fsharp\typecheck\sigs\neg117.bsl] and [D:\a\1\s\tests\fsharp\typecheck\sigs\neg117.err]
line 10
```

**Comment by abelbraaksma** [comment]
The relevant part of the above changed message is:

```diff
- when (Neg117.TargetB.TargetB or  ^f) : (static member Transform :  ^f * Neg117.TargetB.TargetB -> Neg117.TargetB.TransformerKind ->  ^f)
+ when (Neg117.TargetB.TargetB or  ^f) : (static member Transform :  ^f * Neg117.TargetB.TargetB -> (Neg117.TargetB.TransformerKind ->  ^f)) 
```

Note the extra parens that now correctly indicate a function is returned. It now mimics the tooltip:

image

File: src/fsharp/NicePrint.fs (hunk 1 above)
Comment (abelbraaksma) [reply src/fsharp/NicePrint.fs:895]: Fixed in https://github.com/dotnet/fsharp/pull/9728/commits/6c2b3e67d1cf9f0f3b39572940d8e0293d750469 by changing `layoutReturnType` to be in-scope for this line.

**Comment by abelbraaksma** [comment]
Interesting, with the 2 years worth of merging, instead of 3x a version of the FS0001 type error, it now has 1x FS0001 and 2x FS0043, which reminded me of this old issue: https://github.com/dotnet/fsharp/issues/3508.

I'm however at a loss why the error is mentioned 3x in the `*.bsl` file. Whenever I run the code in FSI I get the first error only once. With `fsc.exe` I get it three times:

image

(the 3x reporting was already like this when the PR was first created, except that the error numbers are different now, and this is not related to this PR per se, this PR only fixes the correct signature for the overloads, or missing constraints).

**Comment by abelbraaksma** [comment]
Green, @cartermp, this is ready.

@realvictorprm, do you want to review this? Or you think it's good as is?
